### PR TITLE
[SYCL] Add tests for native math extension

### PR DESCRIPTION
**Pull request by pgorlani**
This patch adds tests for https://github.com/intel/llvm/pull/5747

### CODE REVIEW

File: SYCL/DeviceLib/built-ins/ext_native_math.cpp
Comment (gmlueck) [inline SYCL/DeviceLib/built-ins/ext_native_math.cpp]: Presumably, the `float` and `float{2,3,4,8,16}` tests would still pass, though, right?  Rather than disabling this test for the CPU platform, why not add a runtime check to see if the device has `aspect::fp16`, and then only disable the `half` tests if the device doesn't have that aspect.

Comment (pgorlani) [reply SYCL/DeviceLib/built-ins/ext_native_math.cpp]: Yes, the test passes for floats.

Good point! In order to implement your suggestion, the test need to be compiled with ```-fsycl-device-code-split=per_kernel```, otherwise it will fail at runtime due to unimplemented half functions.

Is it fine?

Comment (gmlueck) [reply SYCL/DeviceLib/built-ins/ext_native_math.cpp]: Yes, it's fine to add `-fsycl-device-code-split=per_kernel` for now.  This won't be necessary once we implement the ["optional  device feature"](https://github.com/intel/llvm/blob/sycl/sycl/doc/design/OptionalDeviceFeatures.md) support.  You could include a comment to that effect, so we know to remove the `-fsycl-device-code-split=per_kernel` later.

Comment (pgorlani) [reply SYCL/DeviceLib/built-ins/ext_native_math.cpp]: Thanks for the info @gmlueck.

I included the modifications in the next commit.

**Review by gmlueck — DISMISSED** [review]
LGTM but someone else should review

**Comment by bader** [comment]
> This patch adds tests for [intel/llvm#5747](https://github.com/intel/llvm/pull/5747)

intel/llvm#5747 changes are not covered by existing pre-commit runs, but new test passes, so it's not clear what is the value of this test if it passes w/o changes it supposed to validate. In other words, I expect the test to fail in pre-commit until intel/llvm#5747 is merged.

**Comment by pgorlani** [comment]
Yes, the tests are performed just if the extension is supported, i.e. ``SYCL_EXT_ONEAPI_NATIVE_MATH`` is defined.

https://github.com/intel/llvm-test-suite/blob/0526d4d1cc6ec02106bbb6396ce93ed2503dbca3/SYCL/DeviceLib/built-ins/ext_native_math.cpp#L31

https://github.com/intel/llvm-test-suite/blob/0526d4d1cc6ec02106bbb6396ce93ed2503dbca3/SYCL/DeviceLib/built-ins/ext_native_math.cpp#L49

**Comment by bader** [comment]
I wonder if we can make the status of the test more clear. Right now it report pass even if nothing was tested.
Usually it's being solved by "REQUIRES: <feature>" and llvm-lit reports this test in "Skipped" list if feature is not enabled.
Probably we need to think now to make it easy for our tests to report correct status for the tests covering optional features/extensions.

**Comment by pgorlani** [comment]
> I wonder if we can make the status of the test more clear. Right now it report pass even if nothing was tested.

I added an assert if the extension is not supported, it this case the test is reported as failed.

>Usually it's being solved by "REQUIRES: " and llvm-lit reports this test in "Skipped" list if feature is not enabled.

In that case, we need all extensions listed as features in [SYCL/lit.cfg.py](https://github.com/intel/llvm-test-suite/blob/e8fa63da6f2daa03dc2d52de42581788a1965340/SYCL/lit.cfg.py).

> Probably we need to think now to make it easy for our tests to report correct status for the tests covering optional features/extensions.

Yes, this is an open question.

File: SYCL/DeviceLib/built-ins/ext_native_math.cpp
Comment (gmlueck) [inline SYCL/DeviceLib/built-ins/ext_native_math.cpp:152]: You are missing tests for `marray` in all of these cases.  According to the [SYCL 2020 spec](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#_description_of_the_built_in_types_available_for_sycl_host_and_device), the "genfloatf" and "genfloath" pseudo-types include `marray` with all possible values of *N*.  I'd suggest testing the same set as `vec` (2, 3, 4, 8, 16) and also at least one weird one (e.g. 43).

File: SYCL/DeviceLib/built-ins/ext_native_math.cpp
Comment (gmlueck) [inline SYCL/DeviceLib/built-ins/ext_native_math.cpp]: This only tests the case when all vector elements have the same value.  It would be good to test the case when each element has a different value.  That will also give you better coverage for a variety of input values.

File: SYCL/DeviceLib/built-ins/ext_native_math.cpp
Comment (pgorlani) [reply SYCL/DeviceLib/built-ins/ext_native_math.cpp:152]: Unfortunately, I don't think this is possible. At the moment, all built-ins (not only the ones tested by this patch) don't accept `marray`'s as arguments.

File: SYCL/DeviceLib/built-ins/ext_native_math.cpp
Comment (pgorlani) [reply SYCL/DeviceLib/built-ins/ext_native_math.cpp]: I took an approach similar to the tests present in `SYCL/DeviceLib/built-ins` and I generalized it for considering not only `float2` but also all the possible vector sizes.

Considering different values is possible, but will complicate the code in a way that could potentially bring problems since the native functions don't have strict accuracy constraints, i.e., each value could have a different range of approximation depending on the device.

Comment (gmlueck) [reply SYCL/DeviceLib/built-ins/ext_native_math.cpp]: Your `assert_out_of_bound` checks already test an arbitrary precision range, so it seems like you could use the same technique for other values.  I noticed that the range was pretty big, so presumably any reasonable device would compute a value within that range.

Comment (pgorlani) [reply SYCL/DeviceLib/built-ins/ext_native_math.cpp]: Sure, but it is one single range. So, it will be easier to tweak in case of problems. If every value is different up to 16 ranges need to be defined.

If this is necessary, I can modify the code.

Comment (gmlueck) [reply SYCL/DeviceLib/built-ins/ext_native_math.cpp]: Testing only the case when all elements of the `vec` have the same value seems like a weak test for this feature.  I think it should be expanded to test the case when each element has a different value.

File: SYCL/DeviceLib/built-ins/ext_native_math.cpp
Comment (gmlueck) [reply SYCL/DeviceLib/built-ins/ext_native_math.cpp:152]: Yes, that does seem to be the case.  I created an internal issue to track this missing feature.

File: SYCL/DeviceLib/built-ins/ext_native_math.cpp
Comment (pgorlani) [reply SYCL/DeviceLib/built-ins/ext_native_math.cpp]: Done.

**Review by gmlueck — APPROVED** [review]
Looks good, thanks!

**Comment by pgorlani** [comment]
Would it be possible to merge this PR? 

Thanks!

File: SYCL/DeviceLib/built-ins/ext_native_math.cpp
Comment (JackAKirk) [reply SYCL/DeviceLib/built-ins/ext_native_math.cpp:152]: Hi @gmlueck : a lot of marray support has been added here: https://github.com/intel/llvm/pull/6038

Comment (gmlueck) [reply SYCL/DeviceLib/built-ins/ext_native_math.cpp:152]: Thanks for this!  I added a note to our internal issue with a link to that PR.

**Review by steffenlarsen — APPROVED** [review]
Sorry for the late review, @pgorlani .

**Pull request merged**
